File: lib/lambda/rules/aws_cis_3_9.ts
import { cloudTrailRecord, uguisuRule, detection } from "../models";

// https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#cis-3.9-remediation

export class rule extends uguisuRule {
  constructor() {
    super({
      id: "aws_cis_3.9",
      title: "AWS Config configuration changes",
      description:
        "AWS CIS benchmark 3.9 recommend to ensure a log metric filter and alarm exist for AWS Config configuration changes",
      severity: "medium",
    });
  }

  detect(record: cloudTrailRecord): detection | null {
    const targetEvents: { [key: string]: boolean } = {
      StopConfigurationRecorder: true,
      DeleteDeliveryChannel: true,
      PutDeliveryChannel: true,
      PutConfigurationRecorder: true,
    };

    if (
      record.eventSource === "config.amazonaws.com" &&
      targetEvents[record.eventName] === true
    ) {
      return {
        rule: this,
        event: record,
      };
    }

    return null;
  }
}

File: pkg/rules/aws_cis_3_9.go
package rules

import (
	"github.com/m-mizutani/uguisu/pkg/models"
)

// https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#cis-3.9-remediation
type awsCIS3_9 struct {
	targetEvents map[string]bool
}

func newAwsCIS3_9() models.Rule {
	return &awsCIS3_9{
		targetEvents: map[string]bool{
			"StopConfigurationRecorder": true,
			"DeleteDeliveryChannel":     true,
			"PutDeliveryChannel":        true,
			"PutConfigurationRecorder":  true,
		}}
}

func (x *awsCIS3_9) ID() string                { return "aws_cis_3.9" }
func (x *awsCIS3_9) Title() string             { return "AWS Config configuration changes" }
func (x *awsCIS3_9) Severity() models.Severity { return models.SeverityMedium }
func (x *awsCIS3_9) Description() string {
	return "AWS CIS benchmark 3.9 recommend to ensure a log metric filter and alarm exist for AWS Config configuration changes"
}
func (x *awsCIS3_9) Match(record *models.CloudTrailRecord) bool {
	return x.targetEvents[record.EventName]
}
